app/src/main/java/org/dhis2/usescases/searchTrackEntity/SearchTEIViewModel.kt

```diff
@@ -0,0 +1,268 @@
+package org.dhis2.usescases.searchTrackEntity
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import androidx.paging.PagedList
+import com.mapbox.geojson.FeatureCollection
+import java.util.HashMap
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import org.dhis2.commons.data.SearchTeiModel
+import org.dhis2.commons.data.uids
+import org.dhis2.commons.network.NetworkUtils
+import org.dhis2.data.search.SearchParametersModel
+import org.dhis2.form.model.RowAction
+import org.dhis2.maps.geometry.mapper.featurecollection.MapCoordinateFieldToFeatureCollection
+import org.dhis2.maps.geometry.mapper.featurecollection.MapTeiEventsToFeatureCollection
+import org.dhis2.maps.geometry.mapper.featurecollection.MapTeisToFeatureCollection
+import org.dhis2.maps.mapper.EventToEventUiComponent
+import org.dhis2.maps.utils.DhisMapUtils
+import org.dhis2.usescases.searchTrackEntity.adapters.uids
+import org.hisp.dhis.android.core.program.Program
+
+class SearchTEIViewModel(
+    initialProgramUid: String?,
+    initialQuery: MutableMap<String, String>?,
+    private val searchRepository: SearchRepository,
+    private val mapTeisToFeatureCollection: MapTeisToFeatureCollection,
+    private val mapTeiEventsToFeatureCollection: MapTeiEventsToFeatureCollection,
+    private val mapCoordinateFieldToFeatureCollection: MapCoordinateFieldToFeatureCollection,
+    private val eventToEventUiComponent: EventToEventUiComponent,
+    private val mapUtils: DhisMapUtils,
+    private val networkUtils: NetworkUtils
+) : ViewModel() {
+
+    val queryData = mutableMapOf<String, String>().apply {
+        initialQuery?.let { putAll(it) }
+    }
+    private val _selectedProgram = MutableLiveData(searchRepository.getProgram(initialProgramUid))
```

Comment:
this call to repository is in the main thread, could this be a problem, should it has a coroutine?